### PR TITLE
feat: clean up output html

### PR DESCRIPTION
**Pull request by sashuk**
[FX-4106]

### Description

This pull request brings the Rich Text Editor output to the same format as the output used in Quill-based editor.

The `cleanupHtmlOutput()` utility was introduced, which takes into account both rules from the Lexical source code and empirical evidence:

- Lexical wraps `<em/>` and `<strong/>` in additional tags for compatibility in https://github.com/facebook/lexical/blob/9b495b8feb100bd535f2a291839cf3a38be08302/packages/lexical/src/nodes/LexicalTextNode.ts#L591, this behavior is removed
- TextNode uses concept of inner tag which is a `<span/>` tag that wraps any text in https://github.com/facebook/lexical/blob/9b495b8feb100bd535f2a291839cf3a38be08302/packages/lexical/src/nodes/LexicalTextNode.ts#L117, this behavior is removed

**Alternative approaches**

- create a `PicassoTextNode` that has a modified `exportDOM()` class method that does not wrap elements in `<b/>`, for example (as it is done in `TextNode` [here](https://github.com/facebook/lexical/blob/2bcf11bec944cd3bd69b7e7b40903e593b716a06/packages/lexical/src/nodes/LexicalTextNode.ts#L598))
  - use node replacement as suggested in https://lexical.dev/docs/concepts/node-replacement, but the editor internals still call the `$createTextNode()` function which creates `TextNode`, not the `PicassoTextNode` – ⚠️ internal library limitations
  - have `registerNodeTransform()` that replaces `TextNodes` with `PicassoTextNode` - ⚠️ the cloning of the element to be replaced requires complex calculations and selection manipulations

### How to test

- Navigate to https://picasso.toptal.net/fx-4106-make-rich-text-editor-output-identical-to-the-old-one-3/ and compare with https://picasso.toptal.net/?path=/story/forms-richtexteditor--richtexteditor

### Development checks

- [N/A] Add changeset according to [guidelines](https://github.com/toptal/picasso/blob/master/docs/contribution/changeset-guidelines.md) (if needed)
- [x] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [N/A] Make sure that additions and changes on design follow [Toptal's BASE design](https://design.toptal.net/), and it's been already discussed with designers at #-base-core
- [N/A] Annotate all `props` in component with documentation
- [N/A] Create `examples` for component
- [N/A] Ensure that deployed demo has expected results and good examples
- [N/A] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).
- [x] Self reviewed
- [x] Covered with tests

> All **_development checks_** should be done and set checked to pass the
> **GitHub Bot: TODOLess** action

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run package:alpha-release` - Release alpha version
- `@toptal-anvil ping reviewers` - Ping FX team for review

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

[FX-4106]: https://toptal-core.atlassian.net/browse/FX-4106?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by TomasSlama — APPROVED** [review]
one nit comment, otherwise looks good

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+              removeChild(child)
+            }
+          }
+        }
+      }
+    })
+  })
+
+  return htmlDoc
+}
```
Comment (dmaklygin) [inline packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts:67]: Let's slightly improve it. This function is too heavy:

```
interface ReplaceCondition {
  child: string;
  nodeToRemove: NodeTypes;
}

const replaceMap: Record<string, ReplaceCondition> = {
  'i': { child: 'em', nodeToRemove: NodeTypes.Parent },
  'b': { child: 'strong', nodeToRemove: NodeTypes.Parent },
  'p': { child: 'span', nodeToRemove: NodeTypes.Child },
  'h3': { child: 'span', nodeToRemove: NodeTypes.Child },
  'li': { child: 'span', nodeToRemove: NodeTypes.Child },
};

export const removeExtraTags = (htmlDoc: Document): Document => {
  Object.keys(replaceMap).forEach(parentTag => {
    const { child, nodeToRemove } = replaceMap[parentTag]
    
    const parentElements = Array.from(htmlDoc.getElementsByTagName(parentTag));

    parentElements.forEach(parentElement => {
      handleParentElement(parentElement, child, nodeToRemove);
    });
  });

  return htmlDoc;
}

const handleParentElement = (parentElement: Element, childTag: string, nodeToRemove: NodeTypes) => {
  const children = Array.from(parentElement.children);

  children.forEach(child => {
    if (child.tagName.toLowerCase() === childTag) {
      removeNode(child, nodeToRemove);
    }
  });
}

const removeNode = (node: Element, nodeType: NodeTypes) => {
  if (nodeType === NodeTypes.Parent) {
    // add the implementation to remove the parent
  }

  if (nodeType === NodeTypes.Child) {
    // add the implementation to remove the child
  }
}
```

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+  for (let index = 0; index < elements.length; index++) {
+    const oldElement = elements[index]
+    const newElement = htmlDoc.createElement(newTag)
+
+    while (oldElement.firstChild) {
+      newElement.appendChild(oldElement.firstChild)
+    }
+
+    oldElement.parentNode!.replaceChild(newElement, oldElement)
```
Comment (dmaklygin) [inline packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts:84]: do we also need to copy attributes?
```
// Preserve attributes
    for (let i = 0; i < oldElement.attributes.length; i++) {
      const { name, value } = oldElement.attributes[i];
      newElement.setAttribute(name, value);
    }
```

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  return htmlDoc
+}
```
Comment (dmaklygin) [inline packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts:88]: btw, both functions are iterate though the elements, that are taken by getElementsByTagName.
we can reuse this code:

```

const handleElements = (
  htmlDoc: Document,
  tag: string,
  callback: (element: Element) => void,
): Document => {
  const elements = Array.from(htmlDoc.getElementsByTagName(tag));
  elements.forEach(callback);
  return htmlDoc;
};

const removeExtraTags = (htmlDoc: Document): Document => {
  Object.keys(replaceMap).forEach(tag => {
    handleElements(htmlDoc, tag, (element) => {
      const condition = replaceMap[tag];
      if (element.tagName.toLowerCase() === condition.child) {
        removeNode(element, condition.nodeToRemove);
      }
    });
  });
  return htmlDoc;
};

const replaceTags = (
  htmlDoc: Document,
  oldTag: string,
  newTag: string,
): Document => {
  handleElements(htmlDoc, oldTag, (oldElement) => {
    const newElement = htmlDoc.createElement(newTag);
    for (let i = 0; i < oldElement.attributes.length; i++) {
      const { name, value } = oldElement.attributes[i];
      newElement.setAttribute(name, value);
    }
    while (oldElement.firstChild) {
      newElement.appendChild(oldElement.firstChild);
    }
    if (oldElement.parentNode) {
      oldElement.parentNode.replaceChild(newElement, oldElement);
    }
  });
  return htmlDoc;
};

```

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.test.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import { cleanupHtmlOutput } from './cleanupHtmlOutput'
+
+// Test cases structure:
+// removeWhat (string), removeFrom (string), input (string), expected (string)
+const tests = [
```
Comment (OleksandrNechai) [inline packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.test.ts:5]: As comments say these are testCases, not tests
```suggestion
const testCases = [
```

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  while (element.firstChild) {
+    if (element.parentNode) {
+      element.parentNode.insertBefore(element.firstChild, element)
+    }
+  }
+
+  element.parentNode!.removeChild(element)
+}
+
+export const removeExtraTags = (htmlDoc: Document): Document => {
```
Comment (OleksandrNechai) [inline packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts:25]: This function is not used anywhere else, maybe no need to export?

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts (hunk 2 above)
Comment (sashuk) [reply packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts:84]: @dmaklygin I think we do not need this as only the italic tag is replaced, so the function is now called `replaceItalicTag()` to avoid its potential misuse

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts (hunk 5 above)
Comment (sashuk) [reply packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts]: @OleksandrNechai Fixed, thanks!

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.test.ts (hunk 4 above)
Comment (sashuk) [reply packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.test.ts]: @OleksandrNechai Fixed, thanks!

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.test.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  ],
+  ['<span>', '<h3> > <span>', '<h3><span>test</span></h3>', '<h3>test</h3>'],
+  ['<span>', '<p> > <span>', '<p><span>test</span></p>', '<p>test</p>'],
+  [
+    'multiple <span> tags',
+    '<p> > <span>',
+    '<p><span>italic </span><i><em>text</em></i><span> test</span></p>',
+    '<p>italic <em>text</em> test</p>',
+  ],
+]
```
Comment (OleksandrNechai) [inline packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.test.ts:26]: I also find implementation a bit too complext to my teste. At first glance it seems above test cases could be done by `str.replace('<i><em>', '<em>').replace('</em></i>', '</em>')`.
If it's not possible then maybe we extracting small JQuery library to manipulate html would be appropriate? Like with such API
```
const $ = fx.load('<p><i><em>italic text</em></i></p>');
$('i em').forEach(e => $(e).parent.replaceWith(e))

```

File: packages/picasso-rich-text-editor/src/LexicalEditor/LexicalEditor.tsx [hunk 7]
```diff
@@ -166,7 +166,9 @@ const LexicalEditor = forwardRef<HTMLDivElement, Props>(function LexicalEditor(
           ? ''
           : removeAttributesFromString($generateHtmlFromNodes(editor, null))
 
-        onChange(htmlValue)
+        const cleanedValue = cleanupHtmlOutput(htmlValue)
+
+        onChange(cleanedValue)
```
[the remark this answers is not in the document]
Comment (sashuk) [reply packages/picasso-rich-text-editor/src/LexicalEditor/LexicalEditor.tsx]: @TomasSlama Changes, thanks!

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts (hunk 3 above)
Comment (sashuk) [reply packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.ts:93]: @dmaklygin Modified, thank you for the suggestion!

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.test.ts (hunk 6 above)
Comment (sashuk) [reply packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.test.ts:26]: @OleksandrNechai During XML tags manipulations I think it is always safer to perform manipulations based on the actual document structure, so we do not miss any edge cases with tags.

> If it's not possible then maybe we could extract a small JQuery-like library to manipulate HTML would be appropriate?

Interesting idea, thanks! However, in the scope of the current ticket, it seems to be more efficient to have custom (and still pretty basic) manipulations without involving any third-party libraries.

**Comment by sashuk** [comment]
@toptal-anvil ping reviewers

File: packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.test.ts (hunk 6 above)
Comment (OleksandrNechai) [reply packages/picasso-rich-text-editor/src/LexicalEditor/utils/cleanupHtmlOutput.test.ts:26]: what I meant is to structure the code in a way that JQuery does, not to pull the whole lib :) . Because currently it is a bit hard to understand.
